src/util/rust_util.rs

```diff
@@ -6,3 +6,68 @@ pub const fn min_of_usize(a: usize, b: usize) -> usize {
         a
     }
 }
+
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+/// Increment the atomic number by 1. If the number reaches limit, return true and reset the number.
+/// Otherwise, return false. This function is thread safe so only the first call that makes the number
+/// reach limit will return true.
+pub fn increment_and_check_limit(u: &AtomicUsize, limit: usize) -> bool {
+    let old = u.fetch_add(1, Ordering::SeqCst);
+
+    if old + 1 >= limit {
+        loop {
+            let current = u.load(Ordering::Relaxed);
+            if current < limit {
+                return false;
+            } else if u.compare_exchange(
+                current,
+                current - limit,
+                Ordering::Release,
+                Ordering::Relaxed,
+            ) == Ok(current)
+            {
+                return true;
+            }
+        }
+    }
+    false
```

Comment:
This function performs memory accesses too many times.  It is a unnecessary, and not atomic. I recommend a transaction-style load-update-store loop:

```suggestion
    match u.fetch_update(Ordering::SeqCst, Ordering::Relaxed, |current| {
        if current >= limit {
            None
        } else if current + 1 < limit {
            Some(current + 1)
        } else {
            Some(0)
        }
    }) {
        Err(old) => panic!("Old vlaue already reached limit: {}", old),
        Ok(old) => old + 1 == limit
    }
```

Reply:
I have made the suggested change with only one difference -- the update function does not return `None`. Can you take a look and see if the PR is okay now?